src/libsyntax/tokenstream.rs

```diff
@@ -178,9 +178,11 @@ impl TokenStream {
             while let Some((pos, ts)) = iter.next() {
                 if let Some((_, next)) = iter.peek() {
                     let sp = match (&ts, &next) {
-                        ((TokenTree::Token(_, token::Token::Comma), NonJoint), _) |
-                        (_, (TokenTree::Token(_, token::Token::Comma), NonJoint)) => continue,
-                        ((TokenTree::Token(sp, _), NonJoint), _) => *sp,
+                        (_, (TokenTree::Token(_, token::Token::Comma), _)) => continue,
+                        ((TokenTree::Token(sp, token_left), NonJoint),
+                         (TokenTree::Token(_, token_right), _))
+                        if token_left.is_ident() || token_left.is_lit() &&
+                            token_right.is_ident() || token_right.is_lit() => *sp,
```

Comment:
Looks like the logical operator priority is unintended and some parentheses are missing.

Reply:
The logic is "a literal or ident is followed by a literal or ident. I can add parentheses and or comments to make it explicit, but it should be correct as is.

Reply:
>The logic is "a literal or ident is followed by a literal or ident"

That's what I inferred too, but `id1 || lit1 && id2 || lit2` is `id1 || (lit1 && id2) || lit2` rather than `(id1 || lit1) && (id2 || lit2)`.

Reply:
You're absolutely right. Fixed.